### PR TITLE
Handle S3 virtual host request type

### LINKED ISSUE #2777: Object Store S3 Alibaba Cloud OSS support
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

For supporting Alibaba Cloud OSS, we need to support:

 1. [Virtual-hosted–style requests](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html#virtual-hosted-style-access): bucket name in hostname instead of in the path
 2. [Signature version s3v4](https://docs.aws.amazon.com/AmazonS3/latest/API/bucket-policy-s3-sigv4-conditions.html)

See downstream issue delta-io/delta-rs#835.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by askoa**
# Which issue does this PR close?

Closes #2777 

# Are there any user-facing changes?
Additional input parameter to determine if the virtual host request type should be use for aws S3.

### CODE REVIEW

**Comment by tustvold** [comment]
This could be my mistake, but running the integration tests with this change I get an error

```
called `Result::unwrap()` on an `Err` value: Generic { store: "S3", source: ListRequest { source: Error { retries: 0, message: "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Key>raphael-test-bucket-e27e29e3</Key><RequestId>579ZMFJ7XFMM0CYV</RequestId><HostId>gdpmDPN2NW2o7FBJ6VG8xM4PTSd5xtuVWXlE8WT8LdznRBUAgTB3zlE7KM66YsU3r2ge2jRHCto=</HostId></Error>", source: reqwest::Error { kind: Status(404), url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("raphael-test-bucket-e27e29e3.s3.eu-west-2.amazonaws.com")), port: None, path: "/raphael-test-bucket-e27e29e3", query: Some("list-type=2"), fragment: None } } } } }
```

The bucket called raphael-test-bucket-e27e29e3 does exist in eu-west-2 so I'm not sure what is going on here?

**Comment by askoa** [comment]
> This could be my mistake, but running the integration tests with this change I get an error
> 
> ```
> called `Result::unwrap()` on an `Err` value: Generic { store: "S3", source: ListRequest { source: Error { retries: 0, message: "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Key>raphael-test-bucket-e27e29e3</Key><RequestId>579ZMFJ7XFMM0CYV</RequestId><HostId>gdpmDPN2NW2o7FBJ6VG8xM4PTSd5xtuVWXlE8WT8LdznRBUAgTB3zlE7KM66YsU3r2ge2jRHCto=</HostId></Error>", source: reqwest::Error { kind: Status(404), url: Url { scheme: "https", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("raphael-test-bucket-e27e29e3.s3.eu-west-2.amazonaws.com")), port: None, path: "/raphael-test-bucket-e27e29e3", query: Some("list-type=2"), fragment: None } } } } }
> ```
> 
> The bucket called raphael-test-bucket-e27e29e3 does exist in eu-west-2 so I'm not sure what is going on here?

The error message says `NoSuchKey`. From the error message I see that the string `raphael-test-bucket-e27e29e3` is used both as bucket and key. Are you passing different argument for key? Does that key exist? The format is `https://bucket-name.s3.region-code.amazonaws.com/key-name`. If you are passing a different `key` argument and if that exist then let me know. There might be some problem with the code.

**Comment by tustvold** [comment]
I'm just running the standard integration tests, with the new environment variable set to "true"...

**Comment by askoa** [comment]
> I'm just running the standard integration tests, with the new environment variable set to "true"... Is it possible that the request URI is still including the bucket when it doesn't need to?

Yes. It looks like I have to update the code in multiple places. I'll revise this PR.

**Comment by askoa** [comment]
@tustvold when an endpoint is given, should we assume it to comply with `virtual_hosted_request_style`? More precisely, if the endpoint is given and `virtual_hosted_request_style` is set to true, should we append bucket name at the end?

https://github.com/apache/arrow-rs/blob/7639f28dceb7397a235f3e32dc60fe6ba1fa90da/object_store/src/aws/mod.rs#L420-L428

File: object_store/src/aws/client.rs [hunk 1]
```diff
@@ -209,14 +209,16 @@ pub struct S3Config {
     pub region: String,
     pub endpoint: String,
     pub bucket: String,
+    pub bucket_endpoint: String,
     pub credentials: CredentialProvider,
     pub retry_config: RetryConfig,
     pub allow_http: bool,
+    pub virtual_hosted_request_style: bool,
```
Comment (tustvold) [inline object_store/src/aws/client.rs:216]: Is this still needed on this object?

File: object_store/src/aws/client.rs [hunk 2]
```diff
@@ -354,7 +356,7 @@ impl S3Client {
         token: Option<&str>,
     ) -> Result<(ListResult, Option<String>)> {
         let credential = self.get_credential().await?;
-        let url = format!("{}/{}", self.config.endpoint, self.config.bucket);
+        let url = self.config.bucket_endpoint.clone();
```
Comment (tustvold) [inline object_store/src/aws/client.rs:359]: Why only this and not all the other URLs?

**Comment by tustvold** [comment]
> I am updating the PR assuming if the input endpoint is provided then we assume path style and append bucket at the end

I would expect the following behaviour:

* If an endpoint is provided and virtual_hosted_request_style - assume bucket already present in provided endpoint
* If no endpoint is provided and virtual_hosted_request_style - construct bucket_endpoint with bucket prefix
* If an endpoint is provided and not virtual_hosted_request_style - include bucket in path

File: object_store/src/aws/client.rs (hunk 2 above)
Comment (askoa) [reply object_store/src/aws/client.rs:359]: This is the only url pre constructed. It does not have `key` in it. The ones with `format!` are different compared to this.

File: object_store/src/aws/client.rs (hunk 1 above)
Comment (askoa) [reply object_store/src/aws/client.rs:216]: Will fix this.

**Comment by askoa** [comment]
> I would expect the following behaviour:
> 
> * If an endpoint is provided and virtual_hosted_request_style - assume bucket already present in provided endpoint
> * If no endpoint is provided and virtual_hosted_request_style - construct bucket_endpoint with bucket name prefix
> * If an endpoint is provided and not virtual_hosted_request_style - include bucket in path

@tustvold I updated the code. Can you please take a look?

**Review by tustvold — APPROVED** [review]
Thank you

File: object_store/src/aws/mod.rs [hunk 3]
```diff
@@ -541,14 +562,29 @@ impl AmazonS3Builder {
...
         };
 
-        let endpoint = self
-            .endpoint
-            .unwrap_or_else(|| format!("https://s3.{}.amazonaws.com", region));
+        let endpoint: String;
+        let bucket_endpoint: String;
+
+        //If `endpoint` is provided then its assumed to be consistent with
+        // `virutal_hosted_style_request`. i.e. if `virtual_hosted_request_style` is true then
```
[the remark this answers is not in the document]
Comment (askoa) [reply object_store/src/aws/mod.rs:596]: `virtual_hosted_style_request` directly comes form AWS page defining it. See https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html. I am not inclined to make this change.

File: object_store/src/aws/mod.rs [hunk 4]
```diff
@@ -455,6 +459,23 @@ impl AmazonS3Builder {
...
 
+    /// Sets if virtual hosted style request has to be used.
+    /// If `virtual_hosted_request_style` is :
+    /// * false (default):  Path style request is used
+    /// * true:  Virtual hosted style request is used
+    ///
+    /// If the `endpoint` is provided then it should be
+    /// consistent with `virtual_hosted_request_style`.
+    /// i.e. if `virutal_hosted_request_style` is set to true
+    /// then `endpoint` should have bucket name on it.
```
[the remark this answers is not in the document]
Comment (askoa) [reply object_store/src/aws/mod.rs]: Updated.

**Pull request merged**